packages/flutter/lib/src/rendering/object.dart

```diff
@@ -251,29 +278,23 @@ class PaintingContext {
       ..isAntiAlias = false;
   }
 
-  /// Paint a child with an opacity
+  /// Push an opacity layer.
   ///
-  /// If the child needs compositing, the blending operation will be applied by
-  /// a compositing layer. Otherwise, the blending operation will be applied by
-  /// the canvas.
-  void paintChildWithOpacity(RenderObject child,
-                             Point childPosition,
-                             Rect bounds,
-                             int alpha) {
-    assert(debugCanPaintChild(child));
-    final Offset childOffset = childPosition.toOffset();
-    if (!child.needsCompositing) {
-      canvas.saveLayer(bounds, _getPaintForAlpha(alpha));
-      canvas.translate(childOffset.dx, childOffset.dy);
-      insertChild(child, Offset.zero);
-      canvas.restore();
+  /// This function will call painter synchronously with a painting context that
+  /// will be blended with the given alpha value.
+  void pushOpacity(bool needsCompositing, Offset offset, Rect bounds, int alpha, PaintingContextCallback painter) {
+    if (needsCompositing) {
+      _stopRecordingIfNeeded();
+      OpacityLayer opacityLayer = new OpacityLayer(bounds: bounds, alpha: alpha);
+      _appendLayer(opacityLayer, offset);
+      PaintingContext childContext = new PaintingContext._(opacityLayer, bounds);
+      painter(childContext, Offset.zero);
+      childContext._stopRecordingIfNeeded();
     } else {
-      OpacityLayer paintLayer = new OpacityLayer(
-          offset: childOffset,
-          bounds: bounds,
-          alpha: alpha);
-      _containerLayer.append(paintLayer);
-      compositeChild(child, parentLayer: paintLayer);
+      // TODO(abarth): pushOpacity should require bounds.
```

Comment:
Nothing changed.  We've had this problem for a long time.  I'm just documenting it.